### PR TITLE
Add LoggerMessageAttribute constructor's overloads

### LINKED ISSUE #87254: Add additional LoggerMessageAttribute constructors
By @geeknoid 

- Event id is not an effective concept; customers don't want to use it and don't have anything to do with it. So, we want overloads that don't require that.
- We are trying to shift customers to not use message strings and stick with structured logging. So, we want overloads that don't include a message.
- We routinely find situations where log levels are specified as a logging method parameter rather than in the attribute definition, so we want overloads that don't include a level.

Bottom line is that the lack of these constructors leads to considerably more ceremony for the customers. More stuff to type with no upside to them.

## Proposal 

```diff
public sealed class LoggerMessageAttribute : Attribute
{
    public LoggerMessageAttribute()
    public LoggerMessageAttribute(int eventId, LogLevel level, string message)

+   public LoggerMessageAttribute(LogLevel level, string message)
+   public LoggerMessageAttribute(LogLevel level)
+   public LoggerMessageAttribute(string message)
}
```

## Original description

dotnet/extensions supports a variety of shortcut constructors on LogMethodAttribute. In the past these shortcuts were deemed undesirable. While I don't know the details behind the original decisions, now that we have more history of internal usage and a better track record we should reconsider.

@geeknoid - can you enumerate exactly which APIs you hope to bring over?

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/87254

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @noahfalk @geeknoid

**Review by geeknoid — APPROVED** [review]

**Comment by geeknoid** [comment]
Note that this change might require an update to the dotnet/runtime logging generator (distinct from the dotnet/extensions extended generator)

**Comment by tarekgh** [comment]
> Note that this change might require an update to the dotnet/runtime logging generator (distinct from the dotnet/extensions extended generator)

What do you mean by `require an update to the dotnet/runtime logging generator`? The changes here already include the changes in the runtime generator.

**Comment by geeknoid** [comment]
> > Note that this change might require an update to the dotnet/runtime logging generator (distinct from the dotnet/extensions extended generator)
> 
> What do you mean by `require an update to the dotnet/runtime logging generator`? The changes here already include the changes in the runtime generator.

Duh. Never mind, I spaced out...

**Review by geeknoid — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs [hunk 1]
```diff
@@ -144,11 +144,43 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
                                             }
 
                                             ImmutableArray<TypedConstant> items = attributeData.ConstructorArguments;
-                                            Debug.Assert(items.Length == 3);
 
-                                            eventId = items[0].IsNull ? -1 : (int)GetItem(items[0]);
-                                            level = items[1].IsNull ? null : (int?)GetItem(items[1]);
-                                            message = items[2].IsNull ? string.Empty : (string)GetItem(items[2]);
+                                            switch (items.Length)
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:148]: Do we correctly handle, e.g.
```C#
[LoggerMessage(level: LogLevel.Warning, message: "custom message", eventId: 0)]
```
?

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs [hunk 2]
```diff
@@ -266,7 +298,8 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
                                     }
 
                                     // ensure there are no duplicate event ids.
-                                    if (!eventIds.Add(lm.EventId))
+                                    // LoggerMessageAttribute has constructors that don't take an EventId, we need to exclude the default Id -1 from duplication checks.
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:301]: We also talked about generating an id based on the EventName if no id was provided.  Are we going to do that?

Comment (geeknoid) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:301]: I'm going to do that in the generator in dotnet/extensions. Not sure if its worth doing in the runtime generator.

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:301]: If we're still having to duplicate the generators, I'd like to minimize the differences between them to only what's truly necessary.

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:301]: I can address this. I am not seeing many users providing Event name with logging though.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:148]: I confirmed this work with and without the new changes.

**Comment by geeknoid** [comment]
So @tarekgh, we should make sure both generators do this the same way. So please let me know what hash function you end up using on the event name in order to produce the event id, and I'll use the same thing on my end.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs (hunk 2 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:301]: > So @tarekgh, we should make sure both generators do this the same way. So please let me know what hash function you end up using on the event name in order to produce the event id, and I'll use the same thing on my end.

I am thinking to copy the non-randomized hashing from https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs,807

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:301]: for the record, I opened the PR https://github.com/dotnet/runtime/pull/87892

**Pull request merged**